Added fragmentation metrics & triggers for GC on idle

cedrichansen opened the pull request:
Trigger compaction if memory fragmentation is detected.

- `LargeObjectAllocateStats::getPageAllignedFreeMemory` returns the amount of space that can reused based on the given page length (obtained through extensions->heap->getPageSize()) due to memory fragmentation. The value returned is an estimate.
- The metric added in the function `MM_ParallelGlobalGC::shouldCompactThisCycle` combines the existing dark matter metric (accessed through `MM_MemoryPool::getDarkMatterBytes()`), with the newly created stat (from bullet 1).
- If the fragmented memory (dark matter + unusable page alligned memory) exceeds a certain threshood, (25% of total memory by default), then a compact is triggered.
- Trigger only occurs when `OMR_GC_IDLE_HEAP_MANAGER` is defined, and `IDLE`.
- Threshold can be specified by `-XXgc:gcOnIdleCompactThreshold=` command line option. Entered value should be a percentage.

- Separate metric was also introduced in `MM_ParallelGlobalGC::shouldCompactThisCycle`, to trigger compaction when dark matter ratio alone exceeds a certain threshold (20% by default).
- Threshold can be specified by '-XXgc:darkMatterCompactThreshold` command line option. Entered value should be a percentage.

- Moved location of call to `processLargeAllocateStatsAfterSweep()` to accomodate changes mentioned above.

- Accompanied by changes on the j9 side for parsing command line options
- addresses issue #3797

Signed-off-by: Cedric Hansen <cedric.hansen@ibm.com>

charliegracie commented:
@genie-omr build all

charliegracie commented:
Hi would you be able to remove the 2 .DS_Store files?

amicic commented on gc/base/j9mm.tdf:
```diff
@@ -916,3 +916,5 @@ TraceEvent=Trc_MM_Scavenger_percolate_preventTenureExpand Overhead=1 Level=1 Gro
 
 TraceAssert=Assert_MM_double_map_unreachable noEnv Overhead=1 Level=1 Assert="(false)"
+
+TraceEvent=Trc_ParallelGlobalGC_shouldCompactThisCycle Overhead=1 Level=1 Group=compact Template="Current fragmented ratio: %f  Threshold: %f"
```
be more specific: "Current page granularity fragmented ratio"

cedrichansen replied:
> Hi would you be able to remove the 2 .DS_Store files?

Yes! I will remove them as soon as I update the commit with the changes suggested

amicic commented on gc/stats/LargeObjectAllocateStats.cpp:
```diff
@@ -262,6 +262,33 @@ MM_FreeEntrySizeClassStats::getFreeMemory(const uintptr_t sizeClassSizes[])
 
+uintptr_t 
+MM_FreeEntrySizeClassStats::getPageAllignedFreeMemory(const uintptr_t sizeClassSizes[], uintptr_t pageSize) {
+
```
Aligned is spelled with a single 'l'

amicic commented on gc/stats/LargeObjectAllocateStats.cpp:
```diff
@@ -262,6 +262,33 @@ MM_FreeEntrySizeClassStats::getFreeMemory(const uintptr_t sizeClassSizes[])
+			/* for each size class, find frequent allocation sizes */
+			if (NULL != _frequentAllocationHead) {
+				MM_FreeEntrySizeClassStats::FrequentAllocation *curr = _frequentAllocationHead[sizeClassIndex];
+				while (NULL != curr) {
```
Insert a comment here some thing like: 
We need to find how many pages (of certain size, which are aligned to its size) would fit into free entries. We don't have the exact info about start and end of each free entry but only the low bound size and count of such entries. Hence we use a simple heuristic that freeEntrySize - pageSize of bytes, in average will fit. For example if free entry is 6K and page is 4K, in average 2K will be reusable (50% chances it will be 1 full page, and 50% no page will fit).

cedrichansen commented:
I believe I have made all of the changes. Changed removed the extra 'l' from getPageAllignedFreeMemory throughout all the files, added the comment in LargeObjectAllocateStats.cpp describing the heuristic being used, j9mm.tdf modified print, and the 2 .DS_Store files were deleted. If anything else needs to be changed, or changes didn't go through as expected, please let me know!

charliegracie commented:
@genie-omr build all

charliegracie commented on gc/base/GCExtensionsBase.hpp:
```diff
@@ -366,6 +369,7 @@ class MM_GCExtensionsBase : public MM_BaseVirtual {
 	float heapSizeStartupHintConservativeFactor; /**< Use only a fraction of hints stored in SC */
 	float heapSizeStartupHintWeightNewValue;		/**< Learn slowly by historic averaging of stored hints */	
 
+
```
can you remove this extra blank line?

charliegracie replied:
NVM about this unless @amicic has other changes that he would like to see added.

charliegracie commented on gc/base/GCExtensionsBase.hpp:
```diff
@@ -1124,6 +1129,11 @@ class MM_GCExtensionsBase : public MM_BaseVirtual {
 	}
 
+	MMINLINE float
+	getDarkMatterCompactThreshold(){
```
Can you put a space between `()` and the `{`?

charliegracie replied:
NVM about this unless @amicic has other changes that he would like to see added.

charliegracie commented on gc/base/standard/ParallelGlobalGC.cpp:
```diff
@@ -464,9 +464,6 @@ MM_ParallelGlobalGC::masterThreadGarbageCollect(MM_EnvironmentBase *env, MM_Allo
 	
 	sweep(env, allocDescription, rebuildMarkBits);
 
-	if (_extensions->processLargeAllocateStats) {
```
I believe when you updated to the latest you lost this recent change. Or were you explicitly removing it?

charliegracie replied:
NVM I see that it just moved down below

cedrichansen commented:
Just updated the GCExtensions base file to remove the extra line, and added the space between the ')' and '{' in the getDarkMatterCompactThreshold function declaration

amicic commented:
Looking good to me.

charliegracie commented:
@genie-omr build all

ShishirH commented:
Sorry to interject, but just to confirm, with these changes, I assume we do not need CompactOnIdle any longer? Does that option do anything now?

amicic commented:
@ShishirH We have not done anything about removing 'always CompactOnIdle'. But yes, this work was meant to replace it. Someone from Could team should verify that is an effective alternative, and than remove CompactOnIdle.
